s3update.go

```diff
@@ -130,41 +132,84 @@ func runAutoUpdate(u Updater) error {
 			return err
 		}
 
-		// Move the old version to a backup path that we can recover from
-		// in case the upgrade fails
 		destBackup := dest + ".bak"
-		if _, err := os.Stat(dest); err == nil {
-			os.Rename(dest, destBackup)
+
+		// Create a temp file
+		tempFile, err := ioutil.TempFile("", "s3update_tmp_download")
+		if err != nil {
+			return err
 		}
+		tempFilePath := tempFile.Name()
 
+		// Download to tempFile
 		// Use the same flags that ioutil.WriteFile uses
-		f, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+		f, err := os.OpenFile(tempFile.Name(), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
 		if err != nil {
-			os.Rename(destBackup, dest)
+			_ = os.Remove(tempFile.Name())
+			return err
+		}
+
+		if err := tempFile.Close(); err != nil {
 			return err
 		}
-		defer f.Close()
 
-		fmt.Printf("s3update: downloading new version to %s\n", dest)
 		if _, err := io.Copy(f, progressR); err != nil {
-			os.Rename(destBackup, dest)
 			return err
 		}
-		// The file must be closed already so we can execute it in the next step
-		f.Close()
 
-		// Removing backup
-		os.Remove(destBackup)
+		// Close the response stream
+		if err := resp.Body.Close(); err != nil {
+			return err
+		}
+
+		// The file must be closed so we can execute it in the next step
+		if err := f.Close(); err != nil {
+			return err
+		}
+
+		if err := finalizeUpdate(dest, destBackup, tempFilePath); err != nil {
+			return err
+		}
 
 		fmt.Printf("s3update: updated with success to version %d\nRestarting application\n", remoteVersion)
 
 		// The update completed, we can now restart the application without requiring any user action.
 		if err := syscall.Exec(dest, os.Args, os.Environ()); err != nil {
+			fmt.Println(err)
 			return err
 		}
 
 		os.Exit(0)
 	}
-
 	return nil
 }
+
+func finalizeUpdate(originalFilePath, backupFilePath, tempFilePath string) (err error) {
+	if downloadSucceeded(tempFilePath) {
+		// Backup current binary
+		if _, err = os.Stat(originalFilePath); err == nil {
+			err = os.Rename(originalFilePath, backupFilePath)
```

Comment:
What made you decide to perform a backup?
Could we simply stick to try to rename the temporary downloaded binary to the destination path?
What can fail (IO interruption, disk space, etc) in those call can still fail in subsequent calls so I'm not sure I understand the benefits.